containers/src/Data/IntMap/Internal.hs

```diff
@@ -62,2837 +275,1076 @@
-              | Tip {-# UNPACK #-} !Key a
-              | Nil
-
-type Prefix = Int
-type Mask   = Int
-
-
--- Some stuff from "Data.IntSet.Internal", for 'restrictKeys' and
--- 'withoutKeys' to use.
-type IntSetPrefix = Int
-type IntSetBitMap = Word
-
-bitmapOf :: Int -> IntSetBitMap
-bitmapOf x = shiftLL 1 (x .&. IntSet.suffixBitMask)
-{-# INLINE bitmapOf #-}
-
-{--------------------------------------------------------------------
-  Operators
---------------------------------------------------------------------}
+{-# INLINE boundsDisjoint #-}
+boundsDisjoint :: Bound L -> Bound R -> Bool
+boundsDisjoint (Bound min) (Bound max) = min > max
 
--- | /O(min(n,W))/. Find the value at a key.
--- Calls 'error' when the element can not be found.
---
--- > fromList [(5,'a'), (3,'b')] ! 1    Error: element not in the map
--- > fromList [(5,'a'), (3,'b')] ! 5 == 'a'
+-- Phantom types used to separate the types of left and right nodes.
+-- They are uninhabited simply to ensure that they are only used as type parameters.
+data L
+data R
 
-(!) :: IntMap a -> Key -> a
-(!) m k = find k m
+#if defined(__GLASGOW_HASKELL__)
+-- TODO: If we are relying on GHC features anyway, L and R could be a new kind.
+newtype Bound t = Bound { boundKey :: Key } deriving (Eq, Ord, Show)
 
--- | /O(min(n,W))/. Find the value at a key.
--- Returns 'Nothing' when the element can not be found.
---
--- > fromList [(5,'a'), (3,'b')] !? 1 == Nothing
--- > fromList [(5,'a'), (3,'b')] !? 5 == Just 'a'
---
--- @since 0.5.11
+type family Flipped t
+type instance Flipped L = R
+type instance Flipped R = L
+#else
+-- Without type families, we can't track min vs. max correctly, so we just don't by making that parameter ignored
+type Bound t = Bound_
+newtype Bound_ = Bound { boundKey :: Key } deriving (Eq, Ord, Show)
+-- This, like L and R, is uninhabited to ensure that it is only used as a type parameter
+data Flipped t
+#endif
 
-(!?) :: IntMap a -> Key -> Maybe a
-(!?) m k = lookup k m
+inMinBound :: Key -> Bound L -> Bool
+inMinBound k (Bound min) = k > min
 
--- | Same as 'difference'.
-(\\) :: IntMap a -> IntMap b -> IntMap a
-m1 \\ m2 = difference m1 m2
+inMaxBound :: Key -> Bound R -> Bool
+inMaxBound k (Bound max) = k < max
 
-infixl 9 !?,\\{-This comment teaches CPP correct behaviour -}
+outOfMinBound :: Key -> Bound L -> Bool
+outOfMinBound k (Bound min) = k < min
 
-{--------------------------------------------------------------------
-  Types
---------------------------------------------------------------------}
+outOfMaxBound :: Key -> Bound R -> Bool
+outOfMaxBound k (Bound max) = k > max
 
-instance Monoid (IntMap a) where
-    mempty  = empty
-    mconcat = unions
-#if !(MIN_VERSION_base(4,9,0))
-    mappend = union
-#else
-    mappend = (<>)
+-- | A map of integers to values @a@.
+newtype IntMap a = IntMap (IntMap_ L a) deriving (Eq)
+data IntMap_ t a = NonEmpty {-# UNPACK #-} !(Bound t) a !(Node t a) | Empty deriving (Eq)
```

Comment:
There are a number of functions and helpers that want to return `IntMap_ R` as intermediates for input to, e.g., `binL`. @treeowl was encouraging the use of unboxed sums for intermediates, so I'm not sure if that /has/ to stay, but it is nice to have that symmetry around.

Reply:
Unboxed sums probably aren't best for all intermediate forms, and they're a pain, but we should look at them *at some point* to see if they can help in any key spots.

Reply:
> There are a number of functions and helpers that want to return `IntMap_ R` as intermediates for input to, e.g., `binL`.

Thanks, that should be in the docs.

Regarding the unboxed sums – I think it would be good to do that as late as possible, ideally after the merge, to simplify the review. Right now I believe the priorities should be documentation and correctness.